Sources/ComposableArchitecture/Store.swift

```diff
@@ -388,45 +412,137 @@ public final class Store<State, Action> {
+      Action
+  ) -> Store<PresentationState<ChildState>, PresentationAction<ChildAction>> {
+    self.scope(
+      state: { $0[keyPath: toChildState] },
+      id: nil,
+      action: { fromChildAction($1) },
+      isInvalid: nil,
+      removeDuplicates: { $0.sharesStorage(with: $1) }
+    )
+  }
+
+  @available(
+    iOS, deprecated: 9999, message: "Pass 'state' a key path to presentation state, instead."
+  )
+  @available(
+    macOS, deprecated: 9999, message: "Pass 'state' a key path to presentation state, instead."
+  )
+  @available(
+    tvOS, deprecated: 9999, message: "Pass 'state' a key path to presentation state, instead."
+  )
+  @available(
+    watchOS, deprecated: 9999, message: "Pass 'state' a key path to presentation state, instead."
+  )
   public func scope<ChildState, ChildAction>(
     state toChildState: @escaping (_ state: State) -> PresentationState<ChildState>,
     action fromChildAction: @escaping (_ presentationAction: PresentationAction<ChildAction>) ->
       Action
   ) -> Store<PresentationState<ChildState>, PresentationAction<ChildAction>> {
     self.scope(
       state: toChildState,
-      action: fromChildAction,
+      id: nil,
+      action: { fromChildAction($1) },
+      isInvalid: nil,
       removeDuplicates: { $0.sharesStorage(with: $1) }
     )
   }
 
   func scope<ChildState, ChildAction>(
     state toChildState: @escaping (State) -> ChildState,
-    action fromChildAction: @escaping (ChildAction) -> Action,
+    id: ((State) -> AnyHashable)?,
+    action fromChildAction: @escaping (State, ChildAction) -> Action,
+    isInvalid: ((State) -> Bool)?,
     removeDuplicates isDuplicate: ((ChildState, ChildState) -> Bool)?
   ) -> Store<ChildState, ChildAction> {
     self.threadCheck(status: .scope)
-    return self.reducer.rescope(
-      self,
-      state: toChildState,
-      action: { fromChildAction($1) },
-      removeDuplicates: isDuplicate
-    )
+
+    let initialChildState = toChildState(self.state.value)
+
+    let id = id?(self.state.value)
+    if
+      let id = id,
+      let childStore = self.children[id] as? Store<ChildState, ChildAction>
+    {
+      return childStore
+    }
+    // NB: This strong/weak self dance forces the child to retain the parent when the parent doesn't
+    //     retain the child.
+    let isInvalid = id == nil
+      ? {
+        self._isInvalidated() || isInvalid?(self.state.value) == true
+      }
+      : { [weak self] in
+        guard let self = self else { return true }
+        return self._isInvalidated() || isInvalid?(self.state.value) == true
+      }
+    let fromChildAction = {
+      BindingLocal.isActive && isInvalid() ? nil : fromChildAction($0, $1)
+    }
+    var isSending = false
+    let childStore = Store<ChildState, ChildAction>(
+      initialState: initialChildState
+    ) {
+      Reduce(internal: { [weak self] childState, childAction in
```

Comment:
We should probably explore using a dedicated reducer conformance for scoping (as was done before with `rescope`), though it'd be nice to measure the difference.

Reply:
I could potentially contain some of the strong/weak dance too.